File: ports/go/icons_test.go
package hqtui

// The same cases as packages/hqtui/test/icons.test.ts.

import (
	"sort"
	"testing"
)

func TestIconModesAndAliases(t *testing.T) {
	cases := []struct {
		name string
		mode IconMode
		want string
	}{
		{"mail", IconNerd, "\U000f01f0"},
		{"mail", IconUnicode, "✉"},
		{"mail", IconASCII, "@"},
		{"email", IconASCII, "@"},
		{"twitter", IconASCII, "x"},
		{"no-such-icon", IconUnicode, ""},
	}
	for _, c := range cases {
		if got := IconIn(c.name, c.mode); got != c.want {
			t.Errorf("IconIn(%q, %s) = %q, want %q", c.name, c.mode, got, c.want)
		}
	}
}

func TestIconNerdFallsBackToUnicode(t *testing.T) {
	for _, row := range openIconGlyphs {
		if row[1] == "" {
			if got := IconIn(row[0], IconNerd); got != row[2] {
				t.Fatalf("%s: nerd fallback = %q, want %q", row[0], got, row[2])
			}
			return
		}
	}
	t.Fatal("expected at least one icon without a Nerd glyph")
}

func TestIconModeOrder(t *testing.T) {
	checks := []struct {
		env  Env
		want IconMode
	}{
		{Env{"OPENICON_GLYPHS": "ascii", "NERD_FONT": "1", "LANG": "en_US.UTF-8"}, IconASCII},
		{Env{"HQTUI_ICONS": "nerd"}, IconNerd},
		{Env{"NERD_FONT": "1"}, IconNerd},
		{Env{"LANG": "en_US.UTF-8", "TERM": "xterm-256color"}, IconUnicode},
		{Env{"TERM": "dumb"}, IconASCII},
	}
	for _, c := range checks {
		if got := IconModeIn(c.env, false); got != c.want {
			t.Errorf("IconModeIn(%v) = %s, want %s", c.env, got, c.want)
		}
	}
	SetIconMode(IconASCII)
	defer SetIconMode("")
	if got := IconModeIn(Env{"NERD_FONT": "1"}, false); got != IconASCII {
		t.Errorf("SetIconMode should win, got %s", got)
	}
}

func TestIconTableSortedAndComplete(t *testing.T) {
	names := IconNames()
	if len(names) < 300 {
		t.Fatalf("only %d icons", len(names))
	}
	if !sort.StringsAreSorted(names) {
		t.Fatal("icons_data.go is not sorted by key")
	}
	for _, row := range openIconGlyphs {
		if row[2] == "" || row[3] == "" {
			t.Errorf("%s has an empty fallback", row[0])
		}
	}
}

File: packages/hqtui/test/icons.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  icon, iconGlyphs, iconMode, iconNames, iconPackFrom, openIcon, setIconMode, useIconPack,
} from "../src/icons.ts";
import { OPENICON_GLYPHS } from "../src/icons-data.ts";
import { renderToText } from "../src/testing.ts";

test("the OpenIcon pack is built in and on by default", () => {
  assert.ok(iconNames().length >= 300);
  for (const key of ["mail", "phone", "link", "search", "settings", "terminal", "git-branch", "github", "x", "slack"]) {
    assert.ok(iconGlyphs(key), key);
  }
});

test("each mode draws its own glyph, and aliases find the same icon", () => {
  assert.equal(icon("mail", { mode: "nerd" }), "\u{f01f0}");
  assert.equal(icon("mail", { mode: "unicode" }), "✉");
  assert.equal(icon("mail", { mode: "ascii" }), "@");
  assert.equal(icon("email", { mode: "ascii" }), "@");
  assert.equal(icon("twitter", { mode: "ascii" }), "x");
});

test("an icon Nerd Fonts lacks falls back to Unicode, and an unknown name draws nothing", () => {
  const noNerd = OPENICON_GLYPHS.find(([, nerd]) => nerd === "");
  assert.ok(noNerd, "the set has icons without a Nerd glyph");
  assert.equal(icon(noNerd[0], { mode: "nerd" }), noNerd[2]);
  assert.equal(icon("no-such-icon"), "");
});

test("the mode comes from the app, then the environment, then the terminal", () => {
  assert.equal(iconMode({ OPENICON_GLYPHS: "ascii", NERD_FONT: "1", LANG: "en_US.UTF-8" }), "ascii");
  assert.equal(iconMode({ HQTUI_ICONS: "nerd" }), "nerd");
  assert.equal(iconMode({ NERD_FONT: "1" }), "nerd");
  assert.equal(iconMode({ LANG: "en_US.UTF-8", TERM: "xterm-256color" }), "unicode");
  assert.equal(iconMode({ TERM: "dumb" }), "ascii");
  setIconMode("ascii");
  try {
    assert.equal(iconMode({ NERD_FONT: "1" }), "ascii");
    assert.equal(icon("phone"), "tel");
  } finally {
    setIconMode(undefined);
  }
});

test("a pack can be swapped for another OpenIcon set", () => {
  const custom = iconPackFrom({
    name: "Tiny", version: "1",
    icons: [{ key: "mail", aliases: ["post"], tui: { unicode: "📮", ascii: "M" } }],
  });
  useIconPack(custom);
  try {
    assert.equal(icon("post", { mode: "ascii" }), "M");
    assert.equal(icon("mail", { mode: "nerd" }), "📮");
    assert.equal(icon("phone"), "");
  } finally {
    useIconPack(openIcon);
  }
});

test("icons render inside widgets like any text", () => {
  setIconMode("ascii");
  try {
    const text = renderToText(({ ui }) => { ui.text(`${icon("mail")} Inbox ${icon("terminal")}`); }, { width: 20, height: 1 });
    assert.ok(text.startsWith("@ Inbox >_"), text);
  } finally {
    setIconMode(undefined);
  }
});

test("the generated table is sorted and complete", () => {
  const keys = OPENICON_GLYPHS.map(([k]) => k);
  assert.deepEqual(keys, [...keys].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0)));
  for (const [key, , unicode, ascii] of OPENICON_GLYPHS) {
    assert.ok(unicode.length > 0, key);
    assert.match(ascii, /^[\x21-\x7e]([\x20-\x7e]{0,2}[\x21-\x7e])?$/, key);
  }
});
